### PR TITLE
Add secrets watcher backend for notifying of secret revision changes

### PR DESCRIPTION
**Pull request by wallyworld**
Add backend support for watching changes to secrets.

The watcher looks at the secrets consumers collection - we need to add the latest secret revision to the doc and this is updated when a secret is updated. Extra logic was also needed to handle concurrent updates.

The Get/Save secrets consumer methods were moved off the secrets service interface and onto a new interface which is backed by State. The secrets service interface is purely for managing secrets themselves.

Also a couple of driveby fixes to naming.

## Checklist

- [x] Code style: imports ordered, good names, simple structure, etc
- [x] Comments saying why design decisions were made
- [x] Go unit tests, with comments saying what you're testing
- [x] [Integration tests](https://github.com/juju/juju/tree/develop/tests), with comments saying what you're testing
- ~[ ] [doc.go](https://discourse.charmhub.io/t/readme-in-packages/451) added or updated in changed packages~

## QA steps

Just unit tests for now, until next PR.

### CODE REVIEW

File: state/secrets.go [hunk 1]
```diff
@@ -370,22 +381,47 @@ func (s *secretsStore) ListSecrets(filter SecretsFilter) ([]*secrets.SecretMetad
...
+		return []txn.Op{countOp}, nil
+	}
+	if inc > 0 {
+		incOp, err := nsRefcounts.CreateOrIncRefOp(refCountCollection, uri.ID, inc)
+		if err != nil {
+			return nil, errors.Trace(err)
+		}
+		return []txn.Op{countOp, incOp}, err
+	}
+	incOp := nsRefcounts.JustIncRefOp(refcountsC, uri.ID, inc)
```
Comment (ycliuhw) [inline state/secrets.go:414]: Can `inc` < 0?

**Review by ycliuhw — APPROVED** [review]
lgtm ty

File: state/secrets.go (hunk 1 above)
Comment (wallyworld) [reply state/secrets.go:414]: yeah, we we add support for deletion

**Pull request merged**